### DISCUSSION TITLE
Stuttering audio transmission and bad dynamic range

### DISCUSSION THREAD

**Discussion by semmelrinde** [pschatzmann/ESP32-A2DP#394] [Q&A] [answered]
I have used the A2DP example from Espressif and also yours. Both produce the same errors. 

Source is PC (DELL XPS 9710): The audio stutters since the beginning of the stream and the stream isn't continous. The volume behaves like expected.
Source is Smartphone (Xiaomi POCO X3): The audio doesn't stutter and there is constant audio output. However the first level of volume on my phone (around 10%) is already really loud. The maximum is already at the third setting (30% volume). After that it just stays at the same volume.

Configuration:
The I2S slave is a TAS5806 amplifier from Texas Instruments and I set the bit rate of both the ESP32 and the amp to 16 Bit. on the ESP32 everything else is configured like following:

`
static i2s_config_t i2s_config = {

      .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX),
      
      .sample_rate = 44100, // updated automatically by A2DP
      
      .bits_per_sample = (i2s_bits_per_sample_t)16,
      
      .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
      
      .communication_format = (i2s_comm_format_t) (I2S_COMM_FORMAT_STAND_MSB),
      
      .intr_alloc_flags = 0, // default interrupt priority
      
      .dma_buf_count = 16,
      
      .dma_buf_len = 64,
      
      .use_apll = true,
      
      .tx_desc_auto_clear = true // avoiding noise in case of data unavailability
  };
  `
  
  
 What could cause this issue?

**Comment by pschatzmann** [comment, marked as answer]
I don't have any issues using OS/X and Linux. So I guess you have to investigate your issue on your OS side.
You can try to increase the I2S buffer e.g dma_buf_count 20 and dma_buf_len 512 and hope that this helps.

For the volume it's not clear what mechanism is used. Some phones just send the audio with decreased volume and only if it is supported my A2DP is handling the volume. So you can play with the different Volume control implementations or provide one yourself. See [the Wiki](https://github.com/pschatzmann/ESP32-A2DP/wiki/Volume-Control). 

But first activate the logging and verify that if you change the volume on the phone you get the corresponding messages

I assume that you use the latest ESP32 Arduino version which is 2.0.7. 
I suggest you check in the Boards Manager if you use Arduino. Some users are still on some 1.0.6 release and wonder why there run into issues...

**Reply by semmelrinde** [reply]
Inceasing the buffer didn't help. I use PlatformIO and there use the Arduino 2.0.7 framework. 

How do I activate the logging? I have been searching for this already as the A2DP example of espressive did this automatically.

**Reply by pschatzmann** [reply]
Same thing, and it documented in the Wiki, just set the log level
build_flags = -DCORE_DEBUG_LEVEL=5

**Reply by semmelrinde** [reply]
Thank you, okay that works. But i do not get any notification on the serial monitor when the volume is changed on the phone. Now also the Bluetooth also disconnects when I connect the ESP32 to the PC.... The thing is that with my bluetooth headphones everything works fine with both devices (PC and Smartphone). 
I have read that some other people in the Internet have the same Problem with the a2dp on ESP32 (Volume / Stuttering).

Also I can say that with an STM32 and an external ADC there haven't ever been issues like this. So I am quite sure that those problems aren't related to the I2S part.

In your opinion what should I look for in the Serial monitor (or elsewhere) to further narrow down the problem?

**Reply by semmelrinde** [reply]
I have tried it with an iPhone aswell, and the problem with the volume is also there. I would say it's even worse. There must be a bug in the code. Probably not yours, but in the one provided in the example by Espressif. I assume this as I haven't modified the code.
Where could I search for this Problem?
